source/common/router/config_impl.h

```diff
@@ -149,6 +155,8 @@ class VirtualHostImpl : public VirtualHost {
   const ConfigImpl& global_route_config_; // See note in RouteEntryImplBase::clusterEntry() on why
                                           // raw ref to the top level config is currently safe.
   std::list<Router::HeaderAddition> request_headers_to_add_;
```

Comment:
I jut spent a while looking at this and was kind of confused, but AFAICT, request_headers_to_add_ is not used anywhere in this file, and all of this is driven by the request_headers_parser_ (in all three places). Can you clean this up? I think there is vestigial code in here that is not actually being used? And does Router::HeaderAddition need to be plumbed into request_headers_parser_?

Reply:
Between this and the next comment, I realized I could just make a RouteEntryImplBase::finalizeResponseHeaders and introduce a ResponseHeaderParser helper class. Basically a structure parallel to what's done for request headers. This also lets us unify the configuration-based response header manipulation in one location (vs having the RouteConfiguration response headers handled in Http::ConnectionManagerUtility).

Reply:
yay, +1, thanks